base/MAPL_Resource.F90

```diff
@@ -0,0 +1,343 @@
+#include "MAPL_Exceptions.h"
+#include "MAPL_ErrLog.h"
+#include "unused_dummy.H"
+
+!=============================================================================
+!FPP macros for repeated (type-dependent) code
+
+#ifdef SET_VAL
+#  undef SET_VAL
+#endif
+
+#define SET_VAL(T) \
+type is (T) ;\
+   if (default_is_present .and. .not. label_is_present) then ;\
+      select type(default) ;\
+      type is(T) ;\
+         val = default ;\
+      class default ;\
+         _FAIL("Type of 'default' does not match type of 'val'.") ;\
+      end select ;\
+   else ;\
+      call ESMF_ConfigGetAttribute(config, val, label = actual_label, _RC) ;\
+   end if
+
+#ifdef SET_VALS
+#  undef SET_VALS
+#endif
+
+#define SET_VALS(T) \
+type is (T) ;\
+   if (default_is_present .and. .not. label_is_present) then ;\
+      select type(default) ;\
+      type is(T) ;\
+         vals = default ;\
+      class default ;\
+         _FAIL("Type of 'default' does not match type of 'vals'.") ;\
+      end select ;\
+   else ;\
+      call ESMF_ConfigGetAttribute(config, valuelist = vals, count = count, label = actual_label, _RC) ;\
+   end if
+
+#ifdef SET_STRINGS
+#  undef SET_STRINGS
+#endif
+
+#define SET_STRINGS(T, TS, TF) \
```

Comment:
Macros are confusing enough.   Consider making the template arguments a bit move explicit like "T_fmt" and "T_str"?

And maybe `TO_STRING` as the macro name?  Should certainly be singular, not plural if I'm reading this correctly.

Reply:
Good idea for the names. The macro actually sets two strings, so I'll leave the name as is.

Reply:
Ah, I see.  One solution would be to bring the ultimate `print` statement into the macro.   But we really do want to minimize the amount of logic in our macros.  (This is supposed to be Fortran after all.)

But maybe put both output strings into the macro as with `SET_VAL` to make the action a bit clearer at the point of invocation.

Reply:
I agree. The print block is rather ghastly. I updated Matt's branch for the #1453 issue with better logic. I think it's time for me to manually pull that branch in. Originally, I was planning a second PR for Matt's branch, but the refactoring made the two branches incompatible, so I might as well fold Matt's branch in manually. That should clean up the logic.

Reply:
I'll leave this unresolved until I do that.

Reply:
Merging will be simpler if this PR is merged into `develop` before the PR for issue #1686. Please approve.